activerecord/lib/active_record/attributes.rb

```diff
@@ -205,14 +205,21 @@ module ClassMethods
       # tracking is performed. The methods +changed?+ and +changed_in_place?+
       # will be called from ActiveModel::Dirty. See the documentation for those
       # methods in ActiveModel::Type::Value for more details.
-      def attribute(name, cast_type = nil, **options, &block)
+      def attribute(name, cast_type = nil, **options, &_decorator)
         name = name.to_s
         reload_schema_from_cache
 
-        self.attributes_to_define_after_schema_loads =
-          attributes_to_define_after_schema_loads.merge(
-            name => [cast_type || block, options]
-          )
+        prev_cast_type, prev_options, prev_decorator = attributes_to_define_after_schema_loads[name]
+
+        unless cast_type && prev_cast_type
```

Comment:
I don't understand what this check does?

Reply:
This condition prevents inapplicable options or decorators from being retained when a cast type is overridden.  So if a cast type was previously specified, and a new cast type is specified, only the new options and decorator are used.  Otherwise, we try to retain whatever type, options, and decorator we have.